SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,181 @@
+	}
+
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+	Super::EndPlay(EndPlayReason);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	if (UWorld* World = GetWorld())
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Send a new ping, which will trigger a self-perpetuating sequence via timers.
+		SendNewPing();
+		// Set looping timer to 'tick' this component, it doesn't send any pings but matches the MinPingInterval for passing updates to the owning controller.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	if (UWorld* World = GetWorld())
+	{
+		bIsPingEnabled = false;
+		// Clear the timers.
+		World->GetTimerManager().ClearTimer(PingTickHandle);
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		// Reset ping output value.
+		RoundTripPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	// Pass latest measured ping to owning controller to be processed by PlayerState.
+	OwningController->UpdatePing(RoundTripPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	// Generate a new ping ID.
+	uint16 NewPingID = LastSentPingID + 1;
+	// Send and record new ping ID.
+	SendServerWorkerPingID(NewPingID);
+	LastSentPingID = NewPingID;
+	LastSentPingTimestamp = FPlatformTime::Seconds();
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Set a timeout timer to await a reply.
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::OnPingTimeout, TimeoutLimit, false);
+	}
+}
+
+void USpatialPingComponent::OnPingTimeout()
+{
+	// Update round trip ping to reflect timeout.
+	TimeoutCount++;
+	RoundTripPing = TimeoutLimit * TimeoutCount;
+	// Attempt to send another ping.
+	SendNewPing();
+}
+
+void USpatialPingComponent::OnRep_ReplicatedPingID()
+{
+	// Check that replicated ping ID matches the last sent AND check ping is enabled to catch cases where a value is replicated after being locally disabled.
+	if (ReplicatedPingID == LastSentPingID && bIsPingEnabled)
+	{
+		UWorld* World = GetWorld();
```

Comment:
```suggestion
```

Reply:
I think this one is worth leaving as is. The world reference is used multiple times in that block so assigning it just once there saves additional GetWorld() calls.